### ISSUE TITLE
API construction is too lenient

### ISSUE THREAD

**Issue by Crote** [haskell-servant/servant#659]
All of the following compile, but produce code which is nonsense in practice:

- Static path parts containing slashes:
  ```haskell
  "api/v1/item" :> Get '[JSON] [Item]
  ```
  This should be done as `"api" :> "v1" :> "item" :> ...`, and the resulting API is not functional.

- Case-sensitive headers:
  ```haskell
    Header "Authentication" String :> Header "authentication" String :> "item" :> Get '[JSON] [Item]
  ```
  According to the [RFC](https://tools.ietf.org/html/rfc7230#section-3.2), header names are case sensitive. Servant already handles this correctly by assigning the same value to both if either is set, but allowing both to be defined could theoretically result in hard-to-find bugs, as a client _can_ send both at the same time.

- Duplicate headers:
  ```haskell
    Header "authentication" String :> Header "authentication" String :> "item" :> Get '[JSON] [Item]
  ```
  Obviously, the resulting server will be passed two identical values. This should not be allowed, as some external code which parses this may rely on the uniqueness of headers.

**Comment by phadej** [comment]
there is no way to rule this compilation time, yet we could have `HasValidate` type class, which would catch those (e.g. collect headers, check that path parts don't contain slashes).

... Or maybe that check can be done on `Router`? ping @kosmikus

**Comment by phadej** [comment]
I answer myself: `Router` isn't enough, as it doesn't collect headers (they don't affect routing)

**Comment by phadej** [comment]
Another point came up on IRC: duplicate content types in content-type lists

**Comment by alpmestan** [comment]
@Crote It's unfortunately not doable to rule out _all_ of the nonsensical APIs at compile time. While we can collect headers and compare the name for equality and things like that, there are many things we cannot do at the type-level.

Now, with that out of the way, we could still do something _extremely_ useful: provide an interpretation that scans the APIs for a whole bunch of such nonsensical/dangerous "patterns" and then reports everything it has found in a way or another (console output, HTML, whatever). An API validation interpretation, basically. This could become a somewhat standard "linting" tool that people run on their APIs as part of CI, or when something's going wrong when running the server (think of all those cases where the routing behavior is confusing and you want to find out where your request goes, etc).

I'm hoping this would be able to help in addressing #449 / #451. I have too much to do to start working on that but I'd be glad to help anyone interested in doing that kind of work, if any help is needed.

**Comment by arianvp** [comment]
This is defenitely a wonderful case where  `TypeErrors` could help I guess.   Can we pattern match on type-level strings like we can do in Idris or Agda? then we could just make this a compiler error.

**Comment by alpmestan** [comment]
Yes, there are a few opportunities where we can detect things at the typelevel and give a custom type error explaining what went wrong, but some other things have to happen at the value level.
